src/index.js

```diff
@@ -101,6 +101,10 @@ class Bitswap {
     this._updateReceiveCounters(peerId.toB58String(), block, has)
 
     if (has || !wasWanted) {
+      if (wasWanted) {
```

Comment:
Rather than sending the notifications here, I think maybe it makes more sense to send them after the put has completed

Reply:
This code path doesn't execute the `.put`

Reply:
Ah sorry you're right it doesn't execute `.put()`

So if we're at this line then either
- the block is already in the blockstore
  That means this node must have previously sent a notification when the block was put into the blockstore
- the block was not wanted by this node
  So this node will not put the block into the blockstore, and should not tell anyone else it has the block

Reply:
The third way here is when you’ve asked for and received the block but the blockstore hasn’t finished writing it (.has returns false) when you ask for the block again.

The blockstore finishes writing (so .has now returns true), then a peer gives you the block and you are then here. The block is in the blockstore but also in the want list.

The test I add demonstrates this problem.

Reply:
Just to make sure I understand, the problem occurs when:
1. Client requests block with CID1 from Bitswap
2. Bitswap adds CID1 to the wantlist
3. Bitswap receives block with CID1 from a peer
4. While processing the incoming block:
  - Bitswap removes CID1 from the wantlist
  - Client (again) requests block with CID1 from Bitswap
  - Bitswap (again) adds CID1 to the wantlist
  - Bitswap returns block to first request from client
  - Bitswap fails to respond to the second request from the client

My suggested solution would be to only remove the want from the wantlist once we've sent the notification. That way if a second request for the block comes in, the notifier will answer both requests when it eventually receives the block.

Reply:
I think the problem there is that processing the blocks is async, so if you delay sending the notifications until after processing is done, and delay removing from the wantlist until after notifications are done, there's a window during which duplicate block receives can arrive - when they do the first version of the block is being held in a not-yet-committed-batch by the datastore (so blockstore.has returns false) and is still in the wantlist.

I believe this is why there's a comment in the code to the effect of:

```javascript
// quickly send out cancels, reduces chances of duplicate block receives
```

There's also the another window where the batch of blocks is part way through processing so `blockstore.has` may return true while the block is still in the wantlist.  This is definitely a bug with the blockstore, a batch should be atomic but I don't think it's how it's implemented currently.

A better approach might be to remove a given block from the wantlist as soon as the block is received and before any async work is done, but to still set up internal wantBlock notification listeners until the block has been processed, then once it's been processed, send the notifications and stop setting up any new notification listeners, though I think this PR could be viewed as a tactical fix and that one more strategic.  The code in this module is due a bit of an overhaul, tbh.

Reply:
Ok yeah that sounds like a good solution to me.

I would rather implement the strategic fix now, rather than patching with a tactical fix first then strategic fix later. I'm happy to do that work myself.

What do you think about
- sending notifications after committing blocks as part of this PR
- a separate PR for taking care of wantlist / blockstore synchonization (I can do this one)?